C/Kay Roemer/NET-0.21/TOOL/NETSTAT.8
.TH netstat 8 MiNT-Net
.SH NAME
netstat \- show active network connections
.SH SYNOPSIS
.B "netstat [-f domain]"
.SH DESCRIPTION
.I Netstat
is used to display infomation about active communication
endpoints in either the
.B unix
or
.B inet
communication domain or both of them.
If invoked without arguments it shows only
.B unix
domain sockets.
You can specify the domain to deal with with the -f option.
.sp
For the unix domain the output of netstat consists of one line for
each socket in the system,
which contains the following entries:
.TP 15
.SM "Proto"
The protocol used by the socket, this can be one of:
.RS 15
.TP 15
.SM "unix"
This is the one and only protocol for unix sockets.
.TP
.SM "unknown"
The protocol is unknown, should never happen.
.RE
.TP 15
.SM "Flags"
These are the flags set on the socket. The output for this entry
consists of a space-separated list of tree-letter flag values.
Possible flags are:
.RS 15
.TP 15
.SM "ACC"
The socket is listening for incoming connections.
This applies only to stream sockets.
.TP
.SM "RCV"
The socket is shut down for receiving after a call to
.IR shutdown (3),
ie the socket cannot receive any messages.
.TP
.SM "SND"
The socket is shut down for sending after a call to
.IR shutdown (3),
ie it is impossible to send messages on the socket.
.TP
.SM "MRK"
If out-of-band data is handled inline, "receiving is at mark", ie
all data before the out of band data has been read.
.RE
.TP 15
.SM "State"
This field is the internal state of the socket,
the following values are possible:
.RS 15
.TP 15
.SM "VIRGIN"
The socket has just been created,
but has not been assigned a communication domain and type.
.TP
.SM "UNCONNECTED"
The socket is not connected to a peer.
.TP
.SM "CONNECTING"
The socket is about to connect to another one.
.TP
.SM "CONNECTED"
The socket is connected to another one.
.TP
.SM "DISONNECTING"
The socket is about to finish the connection to its peer.
.TP
.SM "DISCONNECTED"
The socket has finished a connection and is again unconnected.
.TP
.SM "UNKNOWN"
This should never happen.
.RE
.TP 15
.SM "Type"
This is the communication style the socket uses.
Possible values are:
.RS 15
.TP 15
.SM "STREAM"
The socket uses connection-oriented byte stream communication,
just like pipes do.
.TP
.SM "DGRAM"
The socket uses connectionless, unrelyable,
datagram oriented communication.
.TP
.SM "RAW"
This is a raw socket.
.TP
.SM "RDM"
Just like
.SM "DGRAM\c"
, except that you can rely on the packets to be delivered successfully.
.TP
.SM "SEQPACKET"
This is like
.SM "RDM\c"
, except that the receiver is guaranteed to get the packets in the
same order the sender sends them.
.TP
.SM "UNKNOWN"
This type should never happen.
.RE
.TP 15
.SM "Recv-Q"
The size of the next unread message for datagram oriented sockets,
and the number of unread bytes for stream oriented ones.
.TP
.SM "Address"
The local address of the socket, if
.IR bind (3)
was done on it.
This is a filename for the unix domain.
.SH OPTIONS
.TP 15
-f domain
Display information for active sockets in domain
.IR domain ,
overriding the default domain.
.SH FILES
.TP 15
/dev/unix
Unix domain information
.SH BUGS
Only unix domain implemented.
.SH AUTHOR
Kay Roemer <roemer@hera.rbi.informatik.uni-frankfurt.de>


C/Kay Roemer/NET-0.21/TOOL/NETSTAT.C
/*
 *	This is a simple netstat(8). Currently only the -f option with
 *	unix as argument is implemented.
 *
 *	(w) '93, Kay Roemer.
 */

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/un.h>
#include "socket.h"	/* these come from mint-net/include/ */
#include "net.h"

#define UNIX_DEVICE	"/dev/unix"

struct unix_info {
	short		proto;	 /* protcol numer, always 0 */
	short		flags;	 /* socket flags, SO_* */
	short		type;	 /* socket type, SOCK_DGRAM or SOCK_STREAM */
	short		state;	 /* socket state, SS_* */
	short		qlen;	 /* bytes in read buffer */
	short		addrlen; /* addrlen, 0 if no address */
	struct sockaddr_un addr; /* addr, only meaningful if addrlen > 0 */
};


char *
decode_state (state)
	short state;
{
	switch (state) {
	case SS_VIRGIN:
		return "VIRGIN";

	case SS_ISUNCONNECTED:
		return "UNCONNECTED";

	case SS_ISCONNECTING:
		return "CONNECTING";
			
	case SS_ISCONNECTED:
		return "CONNECTED";
			
	case SS_ISDISCONNECTING:
		return "DISCONNECTING";
		
	case SS_ISDISCONNECTED:
		return "DISCONNECTED";

	default:
		return "UNKNOWN";
	}
}

char *
decode_type (type)
	short type;
{
	switch (type) {
	case SOCK_DGRAM:
		return "DGRAM";
		
	case SOCK_STREAM:
		return "STREAM";
		
	case SOCK_RDM:
		return "RDM";
		
	case SOCK_SEQPACKET:
		return "SEQPACKET";
		
	case SOCK_RAW:
		return "RAW";
		
	default:
		return "UNKNOWN";
	}
}

char *
decode_flags (flags)
	short flags;
{
	static char strflags[20];

	strcpy (strflags, "");
	if (flags & SO_ACCEPTCON)
		strcat (strflags, "ACC ");
	if (flags & SO_CANTSNDMORE)
		strcat (strflags, "SND ");
	if (flags & SO_CANTRCVMORE)
		strcat (strflags, "RCV ");
	if (flags & SO_RCVATMARK)
		strcat (strflags, "MRK ");

	return strflags;
}

void
show_unix (void)
{
	struct unix_info info;
	char *proto;
	int fd, r;

	fd = open (UNIX_DEVICE, O_RDONLY);
	if (fd < 0) {
		perror ("netstat: open");
		return;
	}
	puts ("Unix domain communication endpoints:");
	puts ("Proto   Flags            Type      State         Recv-Q Address");
	for (;;) {
		r = read (fd, &info, sizeof (info));
		if (!r) return;
		if (r < 0) {
			perror ("netstat: read");
			return;
		}
		if (r != sizeof (info)) {
			printf ("netstat: could not read next unix info\n");
			return;
		}

		switch (info.proto) {
		case 0:
			proto = "unix";
			break;

		default:
			proto = "unknown";
			break;
		}

		printf ("%-7s %-16s %-9s %-13s %6u %s\n",
			proto,
			decode_flags (info.flags),
			decode_type (info.type),
			decode_state (info.state),
			info.qlen,
			info.addrlen ? info.addr.sun_path : "");
	}
	close (fd);
}

void
usage (void)
{
	puts ("Usage: netstat [options]");
	puts ("Options:");
	puts ("\t -f unix");
} 

int
main (argc, argv)
	int argc;
	char *argv[];
{
	short show_unix_opt = 1, show_usage_opt = 0;
	extern char *optarg;
	int c;

	while ((c = getopt (argc, argv, "f:")) != EOF) {
		switch (c) {
		case 'f':
			if (!strcmp (optarg, "unix")) {
				show_unix_opt = 1;
			} else {
				show_usage_opt = 1;
			}
			break;

		case '?':
			show_usage_opt = 1;
			break;
		}
	}
	if (show_usage_opt) {
		usage ();
		return 0;
	}
	if (show_unix_opt) {	
		show_unix ();
	}
	return 0;
}
